### ISSUE TITLE
3.0: Support Parameterised Test Fixtures

### ISSUE THREAD

**Issue by agray** [xunit/xunit#2150]
Before you file a bug...

* Is this a question?
  Ask it on our discussion area: https://github.com/xunit/xunit/discussions
* Is this a bug with VSTest (aka Visual Studio Test Explorer, `dotnet test`, `dotnet vstest`,
  or `vstest.console.exe`)? Open those issues here:
  https://github.com/xunit/visualstudio.xunit/issues/
* Did you search the docs @ https://xunit.net/ ?
* Did you search the issues list to see if someone already reported it?
  When looking for duplicates, make sure to look through open **and** closed issues.
* Did you create a simple repro for the problem?
  Bugs without repro projects will be automatically closed.

**Comment by bradwilson** [comment]
@agray Can you please edit the body of this issue and describe the feature you're requesting (including proposed usage)? Thanks!

**Comment by agray** [comment]
XUnit needs to support Parameterised Test fixtures that functionally work similarly to NUnit 3 parameterised test fixtures.

**Comment by bartelink** [comment]
In order to assist with bringing this to fruition, it would help if you would take the time to summarise the key aspects that covers in the OP. Perhaps there are some aspects of the NUnit impl you don't like? Perhaps you have insight into the critical bits of the feature that make it compelling to you? In other words, I'm confident you can do better than "want copy of feature X, DM me when done KTHXBYE"

**Comment by bradwilson** [comment]
Closing for age. Please consider re-opening a new issue with a complete feature description. Thanks!

**Comment by agray** [comment]
XUnit should implement this feature https://docs.nunit.org/articles/nunit/writing-tests/attributes/testfixture.html#parameterized-test-fixtures in an XUnit way.

**Comment by agray** [comment]
Thank you for reopening @bradwilson

**Comment by qubitz** [comment]
I don't know what @agray has in mind, but... here's how I could see "parameterized test fixtures" being implemented in a way that improves upon NUnit's implementation:

```diff
+[FixtureInlineData(Platform.Ubuntu)
+[FixtureInlineData(Platform.Windows)
public class FileCreationTests : IClassFixture<MockPlatformEnv>
{
  private MockPlatformEnv env;

  public void FileIsCreated { ... }
  public void FileHasGivenName { ... }

  public FileCreationTests(MockPlatformEnv env)
  {
    this.env = env;
  }
}

public class MockPlatformEnv
{
+  public MockPlatformEnv(Platform platform) { ... }
}
```

Collectively, all the tests within `FileCreationTests` would be first be ran with a shared instance of `MockPlatformEnv` which was constructed with `Platform.Ubuntu`. Then again, all tests would be ran with a shared instance of `MockPlatformEnv` which was constructed with `Platform.Ubuntu`.

> The order in which these sets of tests run isn't of particular importance, imo

**Comment by agray** [comment]
@qubitz.  I was looking at Theory MemberData and Theory ClassData this morning. Unfortunately it is only at the individual test level.  XUnit needs to be able to apply the same idea at the class level so that the IEnumerable set can be looped over for ALL tests rather than just one test.

Am I correct in my understanding that XUnit doesn't yet support this?

**Comment by bradwilson** [comment]
Correct, we do not support class-level data.

**Comment by qubitz** [comment]
yeah there's no support yet, something like what I proposed above would allow for class level data. Looks like we are going in a little bit of a circle here. Would my proposal solve the issue @agray? It builds upon the idea of XUnit's existing [class fixtures](https://xunit.net/docs/shared-context), but does not exist yet.

**Comment by agray** [comment]
@qubitz 
I need [FixtureMemberData(SomeClass)] on the class

Where SomeClass derives from IEnumerable

And XUnit loops over the IEnumerable for ALL facts and Theories in the class.

If SomeClass returns 5 rows and there are 10 Facts then 50 tests should be executed

If SomeClass returns 5 rows and there are 10 Facts And 1 Theory with 10 executions then 100 tests should be executed.

**Comment by qubitz** [comment]
Okay cool, we're on the same page. `FixtureInlineData` was just the shortest example. I'd also like to expand the idea to encompass `MemberData` and `ClassData` concepts. @bradwilson what level of complexity would be involved to implement this feature? Is this within the scope of XUnit's purpose?

**Comment by agray** [comment]
XUnit is supposed to have the weight of Microsoft behind it.
Every other major unit test framework can do this.
This is a massive gap in XUnit.

**Comment by bradwilson** [comment]
>XUnit is supposed to have the weight of Microsoft behind it.

Says who? I work on this, alone, in my free time. I don't get paid for this.

**Comment by bradwilson** [comment]
>This is a massive gap in XUnit.

I'm not saying I will or won't do this feature, but "because everybody else does it" is not an argument that carries much weight with me. In point of fact, xUnit.net was designed to be _explicitly_ different from the existing frameworks.

If everybody else does this, why don't you use everybody else? /shrug

(In case it's not clear: I have no tolerance for passive/aggressive demands on my personal time.)

**Comment by agray** [comment]
Thanks for replying @bradwilson. Much appreciated.

> If everybody else does this, why don't you use everybody else?
I do.  I maintain the [Saucery](https://github.com/Sauceforge/Saucery) open source project.  It is currently implemented against JUnit (SauceryJ) and NUnit (Saucery).  I'd love to be able to implement it against XUnit (SauceryX).  JUnit, NUnit and probably even more so, XUnit are the titans in the unit testing 

> I'm not saying I will or won't do this feature,
Thank you.

> xUnit.net was designed to be explicitly different from the existing frameworks.
And indeed it is. Using language features like constructors rather than [Setup] is a great approach and very clean.

I suppose all I am suggesting is that its a shame such a great framework doesn't yet have class level setup to apply to all tests in the class.

Kind Regards,
Andrew

**Comment by Odonno** [comment]
+1 on this. Let us know if you need help @bradwilson

**Comment by mattleibow** [comment]
Just adding a comment here so I get notifications. I am wanting to use xunit for UI tests with appium. The test code will all be the same, just need away to have multiple tests for each class/method.

**Comment by bartelink** [comment]
(as we're triggering notifications... see the notifications button top right if you ever need this
but you added some context so not necessarily saying you should delete your comment
... but maybe I should/will delte this one after you ack it)

**Comment by mattleibow** [comment]
I started this discussion around this idea and things I can do today. https://github.com/xunit/xunit/discussions/2897

But my comment before adds such minimal extra context so probably not worth keeping. But also, it does add some but I am in bed and deleting is extra work at midnight XD

**Comment by agray** [comment]
> Just adding a comment here so I get notifications. I am wanting to use xunit for UI tests with appium. The test code will all be the same, just need away to have multiple tests for each class/method.

Check out github.com/sauceforge/saucery.

Saucery.XUnit on NuGet.

**Comment by qubitz** [comment]
> Check out github.com/sauceforge/saucery

@agray what you have here is functionally similar to what I was expecting. On [this test class](https://github.com/Sauceforge/Saucery/blob/7bf0f76aeaeaff0e15dbb095430d235148ddc76f/Merlin.XUnit/ClickLinkTests.cs#L9) you inherit from a [base class that has a method](https://github.com/Sauceforge/Saucery/blob/7bf0f76aeaeaff0e15dbb095430d235148ddc76f/Saucery.XUnit/SauceryXBase.cs#L17) that can [initialize a fixture with the specified arguments](https://github.com/Sauceforge/Saucery/blob/7bf0f76aeaeaff0e15dbb095430d235148ddc76f/Merlin.XUnit/ClickLinkTests.cs#L14). 

A drawback of this solution is that each method within the test class is required to reinitialize the fixture which could be computationally expensive. Though I suppose it could be cached and diffed. Additionally, each method within the test class would require the same theory data attribute.

Regardless, thank you for taking the time to implement and share your workaround.

**Comment by StephenCleary** [comment]
Is this still necessary if we can use interfaces with static members?

I'm thinking the test class can implement an interface that has static members which act as parameters to the test fixture, and then the test fixture can access those static members off the interface (i.e., its parameters) in its constructor (or wherever).

**Comment by agray** [comment]
> > Check out github.com/sauceforge/saucery
> 
> [@agray](https://github.com/agray) what you have here is functionally similar to what I was expecting. On [this test class](https://github.com/Sauceforge/Saucery/blob/7bf0f76aeaeaff0e15dbb095430d235148ddc76f/Merlin.XUnit/ClickLinkTests.cs#L9) you inherit from a [base class that has a method](https://github.com/Sauceforge/Saucery/blob/7bf0f76aeaeaff0e15dbb095430d235148ddc76f/Saucery.XUnit/SauceryXBase.cs#L17) that can [initialize a fixture with the specified arguments](https://github.com/Sauceforge/Saucery/blob/7bf0f76aeaeaff0e15dbb095430d235148ddc76f/Merlin.XUnit/ClickLinkTests.cs#L14).
> 
> A drawback of this solution is that each method within the test class is required to reinitialize the fixture which could be computationally expensive. Though I suppose it could be cached and diffed. Additionally, each method within the test class would require the same theory data attribute.
> 
> Regardless, thank you for taking the time to implement and share your workaround.

Thanks @qubitz.

Supports XUnit, NUnit and TUnit.  I also have a JUnit implementation.

The point of Saucery is you want to run your tests on a given set of platforms. Parameterized Test Fixture fits the bill.

Happy for you to submit a PR if you see more efficient way to do it.

**Comment by alex34533-hpinc** [comment]
Hello,

I run into the same issue.

I'm using ICollectionFixture to share an instance of a (aspire) DistributedApplication (which is expensive to setup) for integration tests.

I'd like to be able to run all tests again various instances of the DistributedApplication setup with different configuration (example: multiple database provider).

Today, the only way is to run tests multiple times with different command line arguments, which is not supported by VS test explorer, and complicates CI setup.

Having a way to parametrize Collection (much like Theory) would solve this.

This
```
[Collection("Name")]
public class ClassTests { }
```
could become
```
[Collection("Name", ["param1_value1", "param2_value1", ... ])] // instantiate one context
[Collection("Name", ["param1_value2", "param2_value2", ... ])] // instantiate another context
public class ClassTests { } // all tests run twice again each context
```

Of course the fixture class would need to implement a constructor with the proper number and type of parameter.


Thank you!

**Comment by almostchristian** [comment]
Microsoft has an implementation of a custom framework for xUnit v2.x that does this.

https://github.com/microsoft/fhir-server/tree/main/src/Microsoft.Health.Extensions.Xunit

For example:
```csharp

public enum DbType
{
    SqlServer,
    CosmosDB,
    All = SqlServer | CosmosDB
}

[DbFramework(DbType.All)]
public class TestClass(DbFixture dbFixture)
{
    [Fact]
    /// ...
}

public class DbFixture(DbType dbType) : IAsyncLifetime
{
    ///
}
```

The above code will spawn two test classes, `TestClass (SqlServer)` and `TestClass (CosmosDB)`.

I tried to convert their code to use xunit.v3, but to no success. It seems like when I create a new XunitTestClass, the UniqueID of the test cases it generates is filtered out and are never executed. Does anyone know of a custom implementation of `XunitTestFrameworkDiscoverer` using xunit.v3 that I can follow?

**Comment by bradwilson** [comment]
@almostchristian The more I think about this, I think if this were the problem I were trying to solve, I would use source generation. You could make `TestClass` abstract, use a Property setter instead of a ctor, then source-generate new types (`TestClass_SqlServer` and `TestClass_CosmosDB`) which derive from the abstract `TestCase`, create instances of `DbFixture` with the correct `DbType` in `IAsyncLifetime.InitializeAsync` and clean them up in `IAsyncLifetime.DisposeAsync`.

Really, the only "problem" would be that you'd have discovered tests that didn't have source, so you'd need to run them from Test Explorer (the option to run a test via hotkey in the source code wouldn't be an option).

**Comment by almostchristian** [comment]
@bradwilson You're right about source generation, I was leaning towards that, but the problem is the test class names it would generate. It may not be an issue if the parameter type is an enum, but if the parameter type is a boolean, it would look like `TestClass_true` or if it's a string with special characters.

I already have converted tests that follow the pattern you suggested that can be source generated. I found the pattern easy to use however, that I got lazy about writing the source generator, since I would have to deal with cases where the fixture has more than one parameter.

```csharp
// can be source generated
[Collection(nameof(DbFixture_CosmosDb))]
public class TestClass_CosmosDb(DbFixture_CosmosDbfixture) : TestClassBase(fixture);

[CollectionDefinition(nameof(DbFixture_CosmosDb))]
public class DbFixture_CosmosDb() : DbFixtureBase(DbType.SqlServer), ICollectionFixture<DbFixture_CosmosDb>;

[Collection(nameof(DbFixture_SqlServer))]
public class TestClass_SqlServer(DbFixture_SqlServer fixture) : TestClassBase(fixture);

[CollectionDefinition(nameof(DbFixture_SqlServer))]
public class DbFixture_SqlServer() : DbFixtureBase(DbType.SqlServer), ICollectionFixture<DbFixture_SqlServer>;

[GeneratedTestFixture(DbType.SqlServer)]
[GeneratedTestFixture(DbType.CosmosDB)]
public abstract class TestClassBase(DbFixtureBase fixture)
{
}

public abstract class DbFixtureBase(DbType dbType) : IAsyncLifetime
{
}
```

**Comment by almostchristian** [comment]
If anyone's interested, I created a source generator to approximate the TestFixtureAttribute as in NUnit/TUnit. It can support AssemblyFixtures. In my opinion, this should just be built in.

https://gist.github.com/almostchristian/700d1d4d486d2780cc7452dad0f9e5a3

It's very basic, it doesn't do validation or string parameter sanitation. If you have AssemblyFixtures, they should be the first parameter(s) in the abstract fixture constructor. The fixture should only have one constructor. Optional fixture constructor parameters are supported. No issues with Visual Studio Test Explorer, double clicking on a test navigates to the test code.

Sample usage:

```csharp
public abstract class SerializationTestFixtureBase(bool useLegacySerializer, bool useLegacyDeserializer, PermissiveParsingLevel permissiveParsingLevel)
{
}

[TestFixture(false, false, PermissiveParsingLevel.Strict)]
[TestFixture(false, false, PermissiveParsingLevel.Lenient)]
[TestFixture(false, true, PermissiveParsingLevel.Lenient)]
[TestFixture(true, true, PermissiveParsingLevel.Lenient)]
[TestFixture(true, true, PermissiveParsingLevel.Strict)]
public abstract class SerializationTestsBase(SerializationTestFixtureBase fixture)
{
  [Fact]
  public void Test1() { }
}
```

The above code will create 5 tests and 5 fixture types inheriting from SerializationTestsBase and SerializationTestFixtureBase:
```csharp
[Collection(nameof(SerializationTestFixture_False_False_Strict))]
public partial class SerializationTests_False_False_Strict(SerializationTestFixture_False_False_Strict fixture) : SerializationTestsBase(fixture);

[CollectionDefinition(nameof(SerializationTestFixture_False_False_Strict))]
public partial class SerializationTestFixture_False_False_Strict() : SerializationTestFixtureBase(false, false, PermissiveParsingLevel.Strict), ICollectionFixture<SerializationTestFixture_False_False_Strict>;

[Collection(nameof(SerializationTestFixture_False_False_Lenient))]
public partial class SerializationTests_False_False_Lenient(SerializationTestFixture_False_False_Lenient fixture) : SerializationTestsBase(fixture);

[CollectionDefinition(nameof(SerializationTestFixture_False_False_Lenient))]
public partial class SerializationTestFixture_False_False_Lenient() : SerializationTestFixtureBase(false, false, PermissiveParsingLevel.Lenient), ICollectionFixture<SerializationTestFixture_False_False_Lenient>;

[Collection(nameof(SerializationTestFixture_False_True_Lenient))]
public partial class SerializationTests_False_True_Lenient(SerializationTestFixture_False_True_Lenient fixture) : SerializationTestsBase(fixture);

[CollectionDefinition(nameof(SerializationTestFixture_False_True_Lenient))]
public partial class SerializationTestFixture_False_True_Lenient() : SerializationTestFixtureBase(false, true, PermissiveParsingLevel.Lenient), ICollectionFixture<SerializationTestFixture_False_True_Lenient>;

[Collection(nameof(SerializationTestFixture_True_True_Lenient))]
public partial class SerializationTests_True_True_Lenient(SerializationTestFixture_True_True_Lenient fixture) : SerializationTestsBase(fixture);

[CollectionDefinition(nameof(SerializationTestFixture_True_True_Lenient))]
public partial class SerializationTestFixture_True_True_Lenient() : SerializationTestFixtureBase(true, true, PermissiveParsingLevel.Lenient), ICollectionFixture<SerializationTestFixture_True_True_Lenient>;

[Collection(nameof(SerializationTestFixture_True_True_Strict))]
public partial class SerializationTests_True_True_Strict(SerializationTestFixture_True_True_Strict fixture) : SerializationTestsBase(fixture);

[CollectionDefinition(nameof(SerializationTestFixture_True_True_Strict))]
public partial class SerializationTestFixture_True_True_Strict() : SerializationTestFixtureBase(true, true, PermissiveParsingLevel.Strict), ICollectionFixture<SerializationTestFixture_True_True_Strict>;

```

**Comment by bradwilson** [comment]
> It can support AssemblyFixtures. In my opinion, this should just be built in.

I'm confused by this statement. I'm not familiar with `TestFixture` in NUnit/TUnit, but xUnit.net does already support assembly fixtures. https://xunit.net/docs/shared-context#assembly-fixture

**Comment by almostchristian** [comment]
What I meant was that my source gen test fixture can support xUnit v3 AssemblyFixture if you have them.

Below is a snippet of how I use it in my code.

```csharp
[assembly: AssemblyFixture(typeof(DatabaseContainerFixture))]

public class DatabaseContainerFixture : IAsyncLifetime { .. }

public abstract class CustomAuditStoreFixtureBase(DatabaseContainerFixture dbFixture, bool useController, AuditDataStoreMode mode, bool useV2Schema) : HttpIntegrationFixtureBase("customaudit")
{
}

[TestFixture(true, AuditDataStoreMode.AuditLogger, false)]
[TestFixture(true, AuditDataStoreMode.Handler, false)]
[TestFixture(true, AuditDataStoreMode.AuditLogger, true)]
[TestFixture(true, AuditDataStoreMode.Handler, true)]
[TestFixture(false, AuditDataStoreMode.AuditLogger, false)]
[TestFixture(false, AuditDataStoreMode.Handler, false)]
[TestFixture(true, AuditDataStoreMode.Middleware, false)]
[TestFixture(false, AuditDataStoreMode.Middleware, false)]
public abstract class CustomAuditStoreTestsBase(CustomAuditStoreFixtureBase fixture)
{
  // tests go here
}
```
